feat: Refactor serial and I2C code.

haydenroche5 opened the pull request:
- Get rid of conditionals on transaction manager existence and if serial locking should be used by adding no-op implementations of each.
- Handle serial and I2C locking with decorators, serial_lock and i2c_lock, respectively.
- Add timeout functionality to I2C locking. Prior to this commit, it would spin forever if the lock couldn't be acquired.
- For both serial and I2C, combine common code from Transaction and Command into an internal function, _transact. Locking happens over this function.
- Migrate all freestanding serial functions to methods of OpenSerial.
- Create a _read_byte abstraction for the three platforms we support (CPython, MicroPython, CircuitPython).
- Merge the functionality of _preprocess_req into _prepare_request. These are always used together, so it doesn't really make sense to have them as separate functions. Additionally, have _prepare_request do everything needed to transform the request data for transmission. Namely, encode the request string as UTF-8 bytes. Finally, make _prepare_request a method of Notecard. There's no reason for it to be a free function.
- Rename I2C's _send_payload to _transmit.

m-mcgowan commented on notecard/notecard.py:
```diff
@@ -65,119 +71,70 @@
+        try:
+            ret = fn(self, *args, **kwargs)
+        finally:
+            if use_i2c_lock:
```
Is the conditional here still needed with a NoOpLock?

haydenroche5 replied:
I only implemented a NoOpSerialLock, but I agree a NoOpI2CLock is in order.

m-mcgowan commented on notecard/notecard.py:
```diff
@@ -225,81 +211,112 @@ def SetTransactionPins(self, rtx_pin, ctx_pin):
-        if 'cmd' not in req:
-            raise Exception("Please use 'cmd' instead of 'req'")
+    @serial_lock
+    def _transact(self, req, rsp_expected):
```
Nice. This is how I would have done it ;-)

m-mcgowan commented on notecard/notecard.py:
```diff
@@ -225,81 +211,112 @@ def SetTransactionPins(self, rtx_pin, ctx_pin):
-        else:
-            serialCommand(self.uart, req, self._debug, self._transaction_manager)
+        try:
+            transaction_timeout_secs = 30
```
The transaction duration will need to be computed dynamically, which is fairly easily done for the request data, although how we can do that to account for the variable response size is difficult, e.g. a large `card.binary.get` over 9600 serial can take several minutes. This is why I feel we should be timing out if data isn't flowing, but otherwise use fairly large transaction timeouts for the overall request/response transmission. Alternatively, we could expose the timeout on the Transaction method so the caller can take the responsibility for getting it right.

I understand we're not taking this direction, so I'm just pointing out some reasons why it could be a good idea.

m-mcgowan commented on notecard/notecard.py:
```diff
@@ -225,81 +211,112 @@ def SetTransactionPins(self, rtx_pin, ctx_pin):
-            serialCommand(self.uart, req, self._debug, self._transaction_manager)
+        try:
+            transaction_timeout_secs = 30
+            self._transaction_manager.start(transaction_timeout_secs)
```
Nitpick, not essential. transaction_manager.start() could return a ContextManager which calls `stop()` in `__exit__()` so the code doesn't need a try-finally.  e.g.

```
with self._transaction_manager.start(transaction_timeout_secs):
   ...
```

haydenroche5 replied:
Yeah, I was thinking that too the other day. It's cleaner.

haydenroche5 replied:
I'm going to shelve this one for another day. I've got an idea of how to do it, but I don't have a testbed for it currently set up.

m-mcgowan commented on notecard/notecard.py:
```diff
@@ -225,81 +211,112 @@ def SetTransactionPins(self, rtx_pin, ctx_pin):
+
             # Send a message with the length of the incoming bytes followed
             # by the bytes themselves.
             if use_periphery:
```
There's an opportunity here to also use a strategy for i2c read/write to avoid the conditional.

haydenroche5 commented:
Addressed review feedback with latest commit (minus the context manager for TransactionManager, which I'm shifting to the backlog, for now). Tested locally against Swan + CircuitPython, ESP32 + MicroPython, and Raspberry Pi.
